src/units/abilities.cpp

```diff
@@ -1283,33 +1288,66 @@ static bool overwrite_special_affects(const config& special)
 			i = overwriters.erase(i);
 		} else {
 			++i;
+			num_list += 1;
 		}
 	}
-	if(overwriters.empty()){
-		return input;
+
+	if(num_list > 1){
+		std::vector<double> priorityvec;
+		for(const auto& i : overwriters) {
+			auto overwrite_specials = (*i.ability_cfg).optional_child("overwrite");
+			if(overwrite_specials && !overwrite_specials["priority"].empty()){
+				priorityvec.push_back(overwrite_specials["priority"].to_double(0));
+			}
+		}
+		if(!priorityvec.empty()){
+			priority = *(std::max_element(priorityvec.begin(), priorityvec.end()));
+			for(unit_ability_list::iterator i = overwriters.begin(); i != overwriters.end();) {
+				auto overwrite_specials = (*i->ability_cfg).optional_child("overwrite");
+				if(overwrite_specials && overwrite_specials["priority"].to_double(0) == priority) {
+					++i;
+				} else {
+					i = overwriters.erase(i);
+				}
+			}
+		}
 	}
+}
 
+unit_ability_list attack_type::overwrite_special_checking(const std::string& ability, unit_ability_list input, unit_ability_list overwriters, double priority) const
+{
 	for(const auto& i : overwriters) {
 		bool affect_side = ((*i.ability_cfg)["overwrite_specials"] == "one_side");
+		auto overwrite_specials = (*i.ability_cfg).optional_child("overwrite");
 		utils::erase_if(input, [&](const unit_ability& j) {
-			bool is_overwritable = (is_special || !overwrite_special_affects(*j.ability_cfg));
+			auto has_overwrite_specials = (*j.ability_cfg).optional_child("overwrite");
+			bool prior = (priority > 0) ? (has_overwrite_specials && has_overwrite_specials["priority"].to_double(0) == priority) : true;
+			bool is_overwritable = (overwrite_special_affects(*j.ability_cfg) && !prior) || !overwrite_special_affects(*j.ability_cfg);
 			bool one_side_overwritable = true;
 			if(affect_side && is_overwritable){
-				if(special_active_impl(shared_from_this(), other_attack_, *i.ability_cfg, AFFECT_SELF, ability, "filter_student")){
-					one_side_overwritable = special_active_impl(shared_from_this(), other_attack_, *j.ability_cfg, AFFECT_SELF, ability, filter_self);
+				if(special_affects_self(*i.ability_cfg, is_attacker_)){
+					one_side_overwritable = special_affects_self(*j.ability_cfg, is_attacker_);
 				}
-				else if(special_active_impl(other_attack_, shared_from_this(), *i.ability_cfg, AFFECT_OTHER, ability, "filter_student")){
-					one_side_overwritable = special_active_impl(other_attack_, shared_from_this(), *j.ability_cfg, AFFECT_OTHER, ability, filter_self);
+				else if(special_affects_opponent(*i.ability_cfg, !is_attacker_)){
+					one_side_overwritable = special_affects_opponent(*j.ability_cfg, !is_attacker_);
+				}
+			}
+			bool special_matches = true;
+			if(overwrite_specials){
+				auto overwrite_filter = (*overwrite_specials).optional_child("filter_specials");
+				if(overwrite_filter && is_overwritable && one_side_overwritable){
+					special_matches = self_ ? (*self_).ability_matches_filter((*j.ability_cfg), ability, *overwrite_filter, true) : false;
```

Comment:
If #7745 merges, what happens to `ability` on this line?

Reply:
i repull ability declaration in rebase because i need that, but must be removed in https://github.com/wesnoth/wesnoth/pull/7745 because don't used before what filter implemented

Reply:
ability removed(in filter using was 'cosmetic' only)